src/interned.rs

```diff
@@ -189,185 +284,431 @@ where
-            // it will only happen rarely.
-            self.key_map
-                .hasher()
-                .hash_one(&table.get::<Value<C>>(*id).fields)
-        }) {
-            // Data has been interned by a racing call, use that ID instead
-            Ok(slot) => {
-                let id = unsafe { slot.as_ref().0 };
-                let value = zalsa.table().get::<Value<C>>(id);
-                let index = self.database_key_index(id);
-
-                // Sync the value's revision.
-                if value.last_interned_at.load() < current_revision {
-                    value.last_interned_at.store(current_revision);
-                    zalsa.event(&|| {
-                        Event::new(EventKind::DidReinternValue {
-                            key: index,
-                            revision: current_revision,
-                        })
-                    });
-                }
+        // Fill up the table for the first few revisions without attempting garbage collection.
+        if !self.revision_queue.is_primed() {
+            return self.intern_id_cold(
+                db,
+                key,
+                zalsa,
+                zalsa_local,
+                assemble,
+                shard,
+                shard_index,
+                hash,
+            );
+        }
 
-                let durability = if let Some((_, stamp)) = zalsa_local.active_query() {
-                    // Record the maximum durability across all queries that intern this value.
-                    let previous_durability = value
-                        .durability
-                        .fetch_max(stamp.durability.as_u8(), Ordering::AcqRel);
+        // Otherwise, try to reuse a stale slot.
+        let mut cursor = shard.lru.back_mut();
 
-                    Durability::from_u8(previous_durability).max(stamp.durability)
-                } else {
-                    value.durability()
-                };
+        while let Some(value) = cursor.get() {
+            // SAFETY: We hold the lock for the shard containing the value.
+            let value_shared = unsafe { &mut *value.shared.get() };
 
-                // Record a dependency on this value.
-                zalsa_local.report_tracked_read_simple(index, durability, value.first_interned_at);
+            // The value must not have been read in the current revision to be collected
+            // soundly, but we also do not want to collect values that have been read recently.
+            //
+            // Note that the list is sorted by LRU, so if the tail of the list is not stale, we
+            // will not find any stale slots.
+            if !self.revision_queue.is_stale(value_shared.last_interned_at) {
+                break;
+            }
 
-                id
+            // We can't collect higher durability values until we have a mutable reference to the database.
```

Comment:
Why is that? (I did not read up on the discussions on the PR) Would be good to add that to the comment?